File: WhatIsUpdated/Diff.fs
namespace WhatIsUpdated
open System
open System.Linq
open System.Collections
open System.Collections.Generic

    module Diff = 
        let toDic map list = 
            Enumerable.ToDictionary(list, System.Func<_,_>(map), System.Func<_,_>(id))
        let get (hash:IDictionary<_,_>) key =
            hash.[key]
        
        let changed map existing incoming = 
            let mapWith = List.map map
            let (setE, setI) = (existing |> mapWith |> Set.ofList, incoming |> mapWith |>  Set.ofList)
              
            let (hashE, hashI) = ( toDic map existing, toDic map incoming )
            let toAdd = Set.difference setI setE
            let toDel = Set.difference setE setI
            let toChange = Set.intersect setI setE

            {
                ToBeAdded = toAdd   |> Set.toList |>  List.map (get hashI) ;
                ToBeRemoved = toDel |> Set.toList |>  List.map (get hashE) ; 
                ToChange = toChange |> Set.toList |>  List.map (get hashI)
            }

        let Changed map existing incoming = 
            let m = FSharpFunc.FromConverter(map)
            changed m (existing |> List.ofSeq) (incoming |> List.ofSeq)

File: WhatIsUpdated/Diff.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace WhatIsUpdated
{
    public class Diff
    {
        public static Updated<T> Changed<T, TKey>(Func<T, TKey> getKey, IEnumerable<T> existing, IEnumerable<T> updated)
        {
            var e = existing.ToArray();
            var u = updated.ToArray();
            var dicE = existing.ToDictionary(getKey, Id);
            var dicU = updated.ToDictionary(getKey, Id);

            var setE = new HashSet<TKey>(dicE.Keys);
            var setU = new HashSet<TKey>(dicU.Keys);

            var toAdd = setU.Except(setE).Select(key => dicU[key]);
            var toRemove = setE.Except(setU).Select(key => dicE[key]);
            var toChange = setE.Intersect(setU)
                .Select(key => new MaybeChanged<T>(
                                    updated: dicU[key],
                                    existing: dicE[key]));

            return new Updated<T>(toChange: toChange, toBeAdded: toAdd, toBeRemoved: toRemove);
        }

        private static T Id<T>(T val) => val;
    }
}
